### DISCUSSION TITLE
ATC on grblHAL? Anyone cracked the code yet?

### DISCUSSION THREAD

**Discussion by massiveradiation** [grblHAL/core#577] [Ideas]
Hello,
Has anyone had any success implementing grblHAL with an ATC setup? I have a proper ATC router which requires two solenoids, one for tool holder release and one for seating. The spindle also has two PNP sensors built in for tool holder seated and tool holder missing. I'd also like to have a solenoid blow off the tool setter. So, I'd need three additional outputs for relays to pop the solenoids and two inputs for tool holder in/out safety checking.
The hard part is the programming! I'm not a coder. I see that the good folks at RapidCNC have done this for their custom ATC, but no one seems to have the golden ticket for a pneumatic ATC spindle.
I'd be happy to pay someone to develop this and then it could be offered up to the public. Or, if someone knows where I can find grblHAL firmware for ATC, I'm all ears!
Thank you,
Tyler

**Comment by karoria** [comment]
Hi @massiveradiation 
There are 2 ways to implement ATC in grblHAL:
1. Traditional method using plugin code specific to your application for which wiki is already in place.
2. Using "tc.macro" file and enabling expressions and flow control in compile time option. This method is very simple and requires no coding knowledge. All you need to do is write a macro file just like a gcode program and with some expressions. For details please go through the [discussion here](https://github.com/grblHAL/core/discussions/309)

Once you go through the discussion, I hope you will be able to do it yourself. However there might be practical issues for which you can post here.

You can also read this wiki page about [Expressions and Flow Control in grblHAL](https://github.com/grblHAL/core/wiki/Expressions-and-flow-control).

**Comment by massiveradiation** [comment]
I'm excited to dig in!
Thanks Ravi.

**Comment by karoria** [comment]
Yes and forgot to mention that there are commands to trigger outputs or wait for inputs which you can use inside tc.macro by which you can operate tool clamp/declamp, air blast to remove chips, etc.

**Reply by massiveradiation** [reply]
Hi Ravi,
Is there reference documentation for trigger I/O for tc.macro?
Thanks,
Tyler

**Reply by karoria** [reply]
Refer linuxcnc M codes M62 to M68 for all types of IO.
https://linuxcnc.org/docs/devel/html/gcode/m-code.html

**Reply by karoria** [reply]
These M codes works in a normal program file also. Not macro specific. Keep in mind that turning on digital output immediately skips planner buffer and acts immediately. So wherever sequence is important, use "G04 P0.1" line before that line. It just syncs the execution till that line as it puts a wait of 0.1 second.
Be extreme cautious while testing macros. Initially put all the feedrates and rapids slower to avoid any physical damage. Real work is to make a macro which behaves like what we need.😉

**Reply by terjeio** [reply]
> These M codes works in a normal program file also.

There is a difference though, when reading an input with M66 the result from the last execution is available in [parameter 5399](https://github.com/grblHAL/core/wiki/Expressions-and-flow-control#numbered-parameters) - and this can then be used for macro flow control.

**Reply by massiveradiation** [reply]
Hey folks,
After doing some initial reading on this, I think there is a parameter that I forgot to throw into the mix regarding my requirements. I want to use bCNC as the gcode sender software as I need Z height mapping/autoleveling and I love the way this is setup in bCNC. So, the question is, can this be done on the grblHAL side? Also, does anyone know if bCNC will recognize and know what do do with ATC gcode?
M61Q<num>: Sets the current tool 
T0 or M6T0: Moves to the change position and waits 
M6T<not 0>: Moves to the change location and does nothing else 
M06: Executes a tool change sequence 
etc.

Any guidance is appreciated! Particularly in terms of sample code, macros, instructions, etc.

**Reply by terjeio** [reply]
> Also, does anyone know if bCNC will recognize and know what do do with ATC gcode?

As long as bCNC passes M61, M6 and T commands to the controller without triggering any special action in the sender it should work. A different matter is when the [inbuilt grblHAL tool change functionality](https://github.com/grblHAL/core/wiki/Manual-tool-change-protocol) is used, that requires the controller to recognize and handle the new _Tool_ state.

> T0 or M6T0: Moves to the change position and waits

Currently, changing from a tool > 0 to 0 does not call the tool change code - it just flags "no tool" in the spindle.
With an ATC it should drop the tool in the tool rack and return somewhere (original position) with no tool in the spindle? Or whatever the a tool change macro is set up to do?
A `T` command by itself just informs the controller which tool the next `M6` command should pick up. It can be used to preposition a tool carousel or whatever else is relevant since the _ts.macro_ will be called if present. 

> M6T<not 0>: Moves to the change location and does nothing else

This can be done by a _tc.macro_, just end it with a `M0` command. A cycle start command will continue running the program.

[This is a link](https://rapidchangeatc.com/docs/integration/grblhal/) to a commercial implementation of ATC macros for grblHAL which contains some useful information and links to a set of tool change macros.

**Reply by massiveradiation** [reply]
Here is a link to bCNC github discussion around M6. After reading it, it seems like the capability is in place. [https://github.com/vlachoudis/bCNC/issues/107](url)

**Reply by terjeio** [reply]
> After reading it, it seems like the capability is in place.

That link is about the sender trapping T and M6 commands and sending a tool change command sequence in their place to the controller. If this cannot be switched off then an ATC has to be handled by the sender (in a sender specific macro?), not by the controller.

**Comment by karoria** [comment]
Here is video of my ATC mill with rotary axis running grblHAL.
https://youtu.be/IZoj1mqmCWU?si=nuoyFeowB3bck2d0

**Reply by Switch-Witch** [reply]
Hi, impressive video! Would you please be so kind to share the used Postpro and your code to make the atc run? I have a 4th axis myself and struggling to get it run and I am planning to install a atc spindle too.

**Reply by karoria** [reply]
Hi, I have not used any software for programming. I manually typed the program. If you set rotary in degrees, then you can rotate it at any degree using G0A90 or so. For drilling, I used G73 drill cycle. I will share my tc.macro for ATC soon.

**Reply by Switch-Witch** [reply]
Thx for your reply! I would be very glad if you could share your tc.macro for the ATC

**Comment by massiveradiation** [comment]
Impressive! Bravo!

**Comment by massiveradiation** [comment]
Hey Ravi,
Is there a particular grblHAL board you recommend? I'm looking at Phil's
board as it seems to have the most I/O:
https://www.tindie.com/products/philba/grblhal-bob-unkit-for-teensy-41-t41u5xbb/

Thoughts?

**Comment by karoria** [comment]
> Hey Ravi,
> 
> Is there a particular grblHAL board you recommend? I'm looking at Phil's
> 
> board as it seems to have the most I/O:
> 
> https://www.tindie.com/products/philba/grblhal-bob-unkit-for-teensy-41-t41u5xbb/
> 
> 
> 
> Thoughts?
> 
> 
> 
> On Thu, Sep 5, 2024 at 9:23 AM Terje Io ***@***.***> wrote:
> 
> 
> 
> > These M codes works in a normal program file also.
> 
> >
> 
> > There is a difference though, when reading an input with M66 the result
> 
> > from the last execution is available in parameter 5399
> 
> > <https://github.com/grblHAL/core/wiki/Expressions-and-flow-control#numbered-parameters>
> 
> > - and this can then be used for macro flow control.
> 
> >
> 
> > —
> 
> > Reply to this email directly, view it on GitHub
> 
> > <https://github.com/grblHAL/core/discussions/577#discussioncomment-10557768>,
> 
> > or unsubscribe
> 
> > <https://github.com/notifications/unsubscribe-auth/AGUKYUUQPDWLOJZ3L6I3KSDZVBLNLAVCNFSM6AAAAABNN4EJLWVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTANJVG43TMOA>
> 
> > .
> 
> > You are receiving this because you were mentioned.Message ID:
> 
> > ***@***.***>
> 
> >
> 
> 

Yes. I use my customized board. But phils boards are best as they are very well designed. One factor to consider is, whether you want to send gcode via ethernet cable or usb cable? Phil's boards have both options available. Be sure to check that the sender you want to use supports sending via ethernet, if you choose so. USB is default supported by most senders.

**Comment by massiveradiation** [comment]
Hey folks, 
I've been wrapping my brain around ATC and grblHAL for a few weeks now. I'm not a programmer and often get lost while in this deep dive. The information is a bit difficult to follow since it is a living open source project and is scattered across several forums, etc. I do appreciate all the efforts you folks are doing on this project!

That said, I'd love to see a web builder page for ATC configuration, similar to the grblHAL builder, which would create a tc.macro file. Masso has this built into their controllers. Maybe if you check a box for ATC it brings up a sub menu. when compiled you get both the hex and tc.macro file, which goes on the sdcard as I understand it.

The following are some ideas for potential prompts:

Type of ATC?: 

LINEAR (tool holders in a line either on X or Y axis)
- drop (tool is dropped into holder)
- side (tool is slid into pincher from side)
CAROSEL, etc.
(I don't know much about the needs for this type. Perh)

-Number of tool holders?

TOOL HOLDER COORDINATES for DROP type:

Z tool holder TOP height (position spindle must raise to in order to safely move over tool holders.
Z tool holder drop height (bottom position for tool holder release)

TOOL #1
- X position
- Y position

TOOL #2
- X position
- Y position
ETC.

TLO touch off type (SHARED WITH PROBE NC?)
- 1 sensor (probe touches off to determine tool length)
- INPUT NUMBER?

- 2 sensor TYPE (has 2nd switch in case tool is driven past first touch position)
- INPUT NUMBER?
- XY COORDINATES

AIR BLOW OFF SOLENOID?
ON TIME?

PNEUMATIC CONTROLS:
SOLENOID #1: AIR SEAL (ON WHENEVER SPINDLE IS ON)?
-I/O PIN ASSIGNMENT:
SOLENOID #2: RELEASE/GRIP TOOL HOLDER SOLENOID
- I/O PIN ASSIGNMENT:
- TIME ON:
SOLENOID #3: SEAT TOOL HOLDER
-I/O PIN ASSIGNMENT
TIME ON: ?

HOLD SPINDLE MOVEMENT WHEN SEATING TOOL HOLDER?

SOLENOID #4: BLOW OUT ATC SPINDLE CHAMBER:
I/O PIN ASSIGNMENT
TIME ON: 

ETC:

Any takers on this? I'd be willing to throw some motivation money at it.

**Reply by Switch-Witch** [reply]
I would definetly pay for it too.... 😄

**Comment by karoria** [comment]
Hi @massiveradiation @Switch-Witch 
Here is my tc.macro file for my 6 linear tool magazine machine. There are no checks for actual tool gripping/ungripping or other inputs so be cautious using this. This will give you a good start but consider adding inputs at appropriate locations where checking is must.

```
(debug,Changing tool #<_current_tool> to tool #<_selected_tool>)
; save current position
#100=#<_x>
#101=#<_y>
#102=#<_z>
#103=#<_spindle_on> ;saving spindle status in a variable
#104=#<_flood> ;saving coolant status in a variable

;tool change code here
G0 G53 Z0
M5 ;stop spindle
M9 ;stop coolant

;program for tool length reference if machine hasn't performed it or undergone soft reset
O301 if [#5070 NE 1]
G0 G53 X-5 Y-28; go to touch off probe coordinates
G0 G53 Z-71 ; rapid down till a safe distance for sensing
G91 G38.2 Z-60 F1000 ; seek down to Z touch off with speed
G91 G0 Z3 ; bounce back 3mm, still in G91
G91 G38.2 Z-3 F100 ; precise touch off for getting #5063 value written in machine memory
G90
G04 P0.1
(debug,Z referenced at #5063)
G0 G53 Z0
O301 endif

#107=#5063 ; save last touch off value in a variable

O101 if [#<_current_tool> EQ 1]
G0 G53 X-287.5 Y-32 ; move to tool rack position #1
O101 elseif [#<_current_tool> EQ 2]
G0 G53 X-241.5 Y-32 ; move to tool rack position #2
O101 elseif [#<_current_tool> EQ 3]
G0 G53 X-195.3 Y-32 ; move to tool rack position #3
O101 elseif [#<_current_tool> EQ 4]
G0 G53 X-149.1 Y-32 ; move to tool rack position #4
O101 elseif [#<_current_tool> EQ 5]
G0 G53 X-102.8 Y-32 ; move to tool rack position #5
O101 elseif [#<_current_tool> EQ 6]
G0 G53 X-56.4 Y-32 ; move to tool rack position #6
O101 endif

O301 if [#<_current_tool> LE 5]
G0 G53 Z-121.9
G0 G53 Y-8.4
O301 elseif [#<_current_tool> GT 5]
G0 G53 Z-120.8
G0 G53 Y-8.4
O301 endif


G04 P0.1 ; pause for syncing immediate output on/off
M64 P1 ; release current tool
G04 P1 ; pause for 1 second so that pneumatics come to final stroke position
G0 G53 Z-50 ; z axis up after tool release

O201 if [#<_selected_tool> EQ 1]
G0 G53 X-287.5 ; move to tool rack position #1
O201 elseif [#<_selected_tool> EQ 2]
G0 G53 X-241.5 ; move to tool rack position #2
O201 elseif [#<_selected_tool> EQ 3]
G0 G53 X-195.3 ; move to tool rack position #3
O201 elseif [#<_selected_tool> EQ 4]
G0 G53 X-149.1 ; move to tool rack position #4
O201 elseif [#<_selected_tool> EQ 5]
G0 G53 X-102.8 ; move to tool rack position #5
O201 elseif [#<_selected_tool> EQ 6]
G0 G53 X-56.4 ; move to tool rack position #6
O201 endif

G04 P0.1 ; pause for syncing immediate output on/off
M64 P2 ; air blow for cleaning on
G0 G53 Z-117

O401 if [#<_selected_tool> LE 5]
G1 G53 Z-121.9 F2000
O401 elseif [#<_selected_tool> GT 5]
G1 G53 Z-120.8 F2000
O401 endif


G04 P0.1 ; pause for syncing immediate output on/off
M65 P1 ; grip new tool

G04 P1 ; pause for 1 second to settle the pneumatics
M61Q#<_selected_tool> ; set new tool as current
G0 G53 Y-32.0
G04 P0.2
M65 P2 ; air blow off

G0 G53 Z0
; touch off probing
G0 G53 X-5 Y-28 ; go to touch off probe coordinates
G0 G53 Z-71 ; rapid down till a safe distance for sensing
G91 G38.2 Z-60 F1000 ; seek down to Z touch off with speed
G91 G0 Z3 ; bounce back 3mm, still in G91
G91 G38.2 Z-3 F100 ; precise touch off
G90
G04 P0.1
(debug,Z Probed at #5063)
G0 Z[#5063]
G10 L20 P0 Z[#107]
G0 G53 Z0
; return to original position
G0 X#100 Y#101
O202 if [#103 EQ 1] ; restore spindle status
M3
O202 endif
O203 if [#104 EQ 1] ; restore coolant status
M8
O203 endif
G0 Z#102
M99
```

**Comment by massiveradiation** [comment]
Thank you so much for this! I've been sidelined from this project, but
intend to jump back in.

**Comment by massiveradiation** [comment]
Hey Ravi,
I'm getting ready to dig back into this ATC project with FluidNC. Do you
know what commands I would use to wait for inputs. The ATC I have has two
built in PNP sensors which indicate when I tool has been dropped and when a
tool is fully seated.
Cheers,
Tyler

**Reply by terjeio** [reply]
[This](https://github.com/grblHAL/core/discussions/577#discussioncomment-10557768).

**Comment by massiveradiation** [comment]
Ok thanks,
I didn't know if there was a difference between grblHal and FluidNC in
regards to this.
Thanks!

**Reply by terjeio** [reply]
Well, I do not know if FluidNC has only partial functionality for M66 implemented. grblHAL supports wait for triggers and timeout if the underlying pins are interrupt capable.

**Comment by massiveradiation** [comment]
I'll dig in and report back.

**Comment by massiveradiation** [comment]
Would you mind taking a look at this brief description of FluidNC M66 and
tell me what you think? Since I'm new to this, I don't know if the question
has been answered. It seems like it to me, but Bart indicates a "partial
implementation."

http://wiki.fluidnc.com/en/features/supported_gcodes#m66-read-user-input

Thank you!

**Reply by terjeio** [reply]
"Only the immediate mode 0 is supported."

Which means there is no wait/timeout for the [requested event](https://www.linuxcnc.org/docs/html/gcode/m-code.html#mcode:m66) to take place...

**Comment by massiveradiation** [comment]
Ah, So, FluidNC will not support this currently? I wonder why grblHAL can
do it, but FluidNC can't. Same backbone, no?
Thanks!

**Reply by terjeio** [reply]
Same MCU, same SDK - on top on that not the same except some shared inheritance from the legacy Grbl project (which did not support M66). So it can be done, unless their current codebase makes it complicated/impossible.

**Comment by massiveradiation** [comment]
Thanks for the help. I'm over on the FluidNC discord trying to find out
what's possible. I'll report back.
